#! /bin/sh -e
# 10_newpatch.dpatch by  <schizo@debian.org>
# updated for 4.0 by Matthias Klose <doko@debian.org>
#
# DP: Enable biarch support for the 32bit sparc compiler

if [ $# -eq 3 ] && [ "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
	patch $pdir -f --no-backup-if-mismatch -p0 < $0
        cd ${dir}libcpp && autoconf
	;;
    -unpatch)
	patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        cd ${dir}libcpp && autoconf
	;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1;;
esac

exit 0
				    
--- gcc.old/config.gcc	2004-10-22 19:05:48.000000000 +0200
+++ gcc/config.gcc	2004-11-13 10:06:28.000000000 +0100
@@ -1854,8 +1854,18 @@
 	use_fixproto=yes
 	;;
 sparc-*-linux*)		# SPARC's running GNU/Linux, libc6
+	# If cpu is specified, assume we want a 32/64 compiler
+	if test x$with_cpu = x; then
 	tm_file="${tm_file} dbxelf.h elfos.h svr4.h sparc/sysv4.h sparc/linux.h"
 	tmake_file="${tmake_file} sparc/t-crtfm"
+	else
+	tm_file="sparc/biarch64.h ${tm_file} dbxelf.h elfos.h svr4.h sparc/sysv4.h sparc/linux64.h"
+	tmake_file="${tmake_file} sparc/t-linux64 sparc/t-crtfm"
+	float_format=sparc
+	need_64bit_hwint=yes
+	fi
+#	tm_file="${tm_file} dbxelf.h elfos.h svr4.h sparc/sysv4.h sparc/linux.h"
+#	tmake_file="${tmake_file} sparc/t-crtfm"
 	;;
 sparc-*-rtems*)
 	tm_file="${tm_file} dbxelf.h elfos.h svr4.h sparc/sysv4.h sol2.h sparc/sol2.h sparc/elf.h sparc/rtemself.h rtems.h"
--- gcc.old/config/sparc/linux64.h	2004-09-08 07:58:31.000000000 +0200
+++ gcc/config/sparc/linux64.h	2004-11-13 10:04:33.000000000 +0100
@@ -51,8 +51,8 @@
    + MASK_STACK_BIAS + MASK_APP_REGS + MASK_FPU + MASK_LONG_DOUBLE_128)
 #endif
 
-#undef ASM_CPU_DEFAULT_SPEC
-#define ASM_CPU_DEFAULT_SPEC "-Av9a"
+#undef ASM_CPU64_DEFAULT_SPEC
+#define ASM_CPU64_DEFAULT_SPEC "-Av9a"
 
 #ifdef SPARC_BI_ARCH
 
--- libcpp/configure.ac~	2004-09-25 09:35:16.000000000 +0200
+++ libcpp/configure.ac	2004-11-13 10:08:41.000000000 +0100
@@ -115,6 +115,7 @@
 	powerpc*-*-* | \
 	rs6000*-*-* | \
 	s390*-*-* | \
+	sparc-*-linux* | \
 	sparc64*-*-* | ultrasparc-*-freebsd* | \
 	sparcv9-*-solaris2* | \
 	sparc-*-solaris2.[789] | sparc-*-solaris2.1[0-9] | \
